src/NovelRT.Interop/LifetimeExtender.h

```diff
@@ -0,0 +1,57 @@
+// Copyright © Matt Jones and Contributors. Licensed under the MIT Licence (MIT). See LICENCE.md in the repository root
+// for more information.
+#ifndef NOVELRT_INTEROP_LIFETIMEEXTENDER_H
+#define NOVELRT_INTEROP_LIFETIMEEXTENDER_H
+#include <memory>
+#include <vector>
+/*
+ * Extends the lifetime of objects contained in a shared_ptr.
+ * This lets the API consumer take ownership of an object without double indirection.
+ */
+namespace NovelRT::Interop::Lifetime
+{
+    template<typename T> [[nodiscard]] std::vector<std::shared_ptr<T>>& AliveObjects() noexcept
+    {
+        static std::vector<std::shared_ptr<T>> aliveObjects{};
+
+        return aliveObjects;
+    }
+
+    template<typename T> void KeepAlive(std::shared_ptr<T>&& object) noexcept
+    {
+        AliveObjects<T>().emplace_back(std::move(object));
+    }
+
+    template<typename T> void KeepAlive(const std::shared_ptr<T>& object) noexcept
+    {
+        AliveObjects<T>().emplace_back(object);
+    }
+
+    template<typename T> [[nodiscard]] std::shared_ptr<T> Find(T* object) noexcept
```

Comment:
Can't this be `const`?

Reply:
`WithPluginProvider` requires a non-const `std::shared_ptr<T>` so I don't think so.

Reply:
uh, incorrect. The function does not mutate `this`. This method can be `const`. I'm pretty sure.
```suggestion
    template<typename T> [[nodiscard]] std::shared_ptr<T> Find(T* object) const noexcept
```

Reply:
But it's a method contained outside of a class, so it's not possible?

From C++ reference:
> cv - const/volatile qualification, only allowed in non-static member function declarations

Reply:
I see. I thought these members were in a class.

How come they aren't?

Reply:
Thought that making a class would be overkill, what's the point of making a fully static class when C++ has functions with namespaces?

Reply:
It's probably my C# brain talking. I'm not sure if there's a benefit.

Reply:
Considering that this appears to be used across a few classes, I think this makes sense, in which case `const` would fall over at compile-time since there's no class for it.

@RubyNova if you're cool with it being outside of a class, then I think this is fine.

Reply:
C API is not my code owner area primarily - this is why I pinged you in. Your say is the final one here I think.